examples/schelling_segregation/src/main.cu

```diff
@@ -0,0 +1,357 @@
+#include <iostream>
+#include <cstdio>
+#include <cstdlib>
+#include <fstream>
+#include <array>
+
+#include "flamegpu/flamegpu.h"
+
+// Configurable properties
+constexpr unsigned int GRID_WIDTH = 100;
+constexpr float THRESHOLD = 0.70;
+
+constexpr unsigned int A = 0;
+constexpr unsigned int B = 1;
+constexpr unsigned int UNOCCUPIED = 2;
+
+// Agents output their type
+FLAMEGPU_AGENT_FUNCTION(output_type, flamegpu::MessageNone, flamegpu::MessageArray2D) {
+    FLAMEGPU->message_out.setVariable<unsigned int>("type", FLAMEGPU->getVariable<unsigned int>("type"));
+    FLAMEGPU->message_out.setIndex(FLAMEGPU->getVariable<unsigned int, 2>("pos", 0), FLAMEGPU->getVariable<unsigned int, 2>("pos", 1));
+    return flamegpu::ALIVE;
+}
+
+// Agents decide whether they are happy or not and whether or not their space is available
+FLAMEGPU_AGENT_FUNCTION(determine_status, flamegpu::MessageArray2D, flamegpu::MessageNone) {
+    const unsigned int my_x = FLAMEGPU->getVariable<unsigned int, 2>("pos", 0);
+    const unsigned int my_y = FLAMEGPU->getVariable<unsigned int, 2>("pos", 1);
+
+    unsigned int same_type_neighbours = 0;
+    unsigned int diff_type_neighbours = 0;
+
+    // Iterate 3x3 Moore neighbourhood (this does not include the central cell)
+    unsigned int my_type = FLAMEGPU->getVariable<unsigned int>("type");
+    for (auto &message : FLAMEGPU->message_in.wrap(my_x, my_y)) {
+        int message_type = message.getVariable<unsigned int>("type");
+        same_type_neighbours += my_type == message_type;
+        diff_type_neighbours += (my_type != message_type) && (message_type != UNOCCUPIED);
+    }
+
+    int isHappy = (static_cast<float>(same_type_neighbours) / (same_type_neighbours + diff_type_neighbours)) > THRESHOLD;
+    FLAMEGPU->setVariable<unsigned int>("happy", isHappy);
+    unsigned int my_next_type = ((my_type != UNOCCUPIED) && isHappy) ? my_type : UNOCCUPIED;
+    FLAMEGPU->setVariable<unsigned int>("next_type", my_next_type);
+    FLAMEGPU->setVariable<unsigned int>("movement_resolved", (my_type == UNOCCUPIED) || isHappy);
+    unsigned int my_availability = (my_type == UNOCCUPIED) || (isHappy == 0);
+    FLAMEGPU->setVariable<unsigned int>("available", my_availability);
+
+    return flamegpu::ALIVE;
+}
+
+FLAMEGPU_AGENT_FUNCTION_CONDITION(is_available) {
+    return FLAMEGPU->getVariable<unsigned int>("available");
+}
+FLAMEGPU_AGENT_FUNCTION(output_available_locations, flamegpu::MessageNone, flamegpu::MessageArray) {
+    FLAMEGPU->message_out.setIndex(FLAMEGPU->getThreadIndex());
```

Comment:
Why is `getThreadIndex()` still required? I doubt Paul will like this being used.

Reply:
I guess it's because you only have a random selection of locations available, which aren't a contiguous block. So you need a way agents can identify valid spots, and this is a fairly simple method.

Not sure of a better approach off the top of my head, maybe one for discussion in the meeting.

Reply:
Having thought about this, I guess the 'proper' way would be for the `spaces_available` property to be binned, and a `spaces_available` brute force message list created.

Spaces available output a message which only contains their ID, Agents then grab the list, check it's length and generate a random number to grab the message at that index. *(Brute force messages don't currently have direct access like an array, so would need iterating, but there's no practical reason it couldn't be added).*

One benefit of this 'proper' version is that it would better facilitate agents with more specific preferences, as the `spaces_available` message could be improved to list any number of features of the space which the agent must approve of. (Although we might want to think about a more generalised version of Spatial messaging for multi-dimensional data if this were to be a real thing).